src/errorpage.cc

```diff
@@ -1345,17 +1355,6 @@ ErrorState::BuildHttpReply()
         rep->body.set(body);
     }
 
-    // Make sure error codes get back to the client side for logging and
-    // error tracking.
-    if (request) {
-        request->error.update(type, detail);
-        request->error.update(SysErrorDetail::NewIfAny(xerrno));
-    } else if (ale) {
-        Error err(type, detail);
-        err.update(SysErrorDetail::NewIfAny(xerrno));
-        ale->updateError(err);
-    }
```

Comment:
Moving this code up (without changes), before the `if (response_)` check, fixes the primary cause of this bug.

The PR also adjusts the ErrorState constructor dedicated to ERR_RELAY_REMOTE responses to get ALE because the moved code updates ALE, even though those updates are probably not related to the current ERR_RELAY_REMOTE code paths. That improvement also proves that this PR did not miss any cases where ERR_RELAY_REMOTE ErrorState was created without ALE.

This messy code needs quite a bit of additional work. We will follow up with at least one (but probably two) PRs that should address out-of-scope problems already marked in ErrorState declaration and new missing error updates problems discovered while working on this PR.